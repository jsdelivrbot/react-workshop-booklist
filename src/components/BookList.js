import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { fetchBooks } from '../actions/index';
import { totalBooks } from '../actions/index';
import { elementsXPage } from '../actions/index';
import { currentPage } from '../actions/index';

class BookList extends Component {

    constructor(props) {
        super(props);
        this.total = 0;
    }

    //Método del ciclo de vida del componente al cargarlo...
    componentDidMount() {
        this.props.fetchBooks();
    }

    //Método del ciclo de vida del componente después de renderizar
    componentDidUpdate() {
        this.props.totalBooks(this.total);
    }

    showCurrentElementsXPage(books) {
        let limitMax = this.props.current_page * this.props.elements_x_page;
        let limitMin = limitMax - this.props.elements_x_page;
        return books.filter((element,index) => index >= limitMin && index < limitMax);
    }

    getFilterBooks() {
        let filterLowerCase = this.props.filter_book.toString().toLowerCase();
        return this.props.books.filter((book) => book.title.toLowerCase().indexOf(filterLowerCase) != -1 );
    }

    render(){

        let books = this.getFilterBooks().map((book) => {
            return <div key={book.id}>{book.title} - {book.author} - {book.isbn}</div>
        });

        this.total = books.length;

        return (
            <div>
                <div>Book List</div>
                <div style={{height:'250px', width:'400px', overflow:'auto', border:'1px solid black'}}>
                {this.showCurrentElementsXPage(books)}
                </div>
            </div>
        )
    }
}

//Mapear estado de aplicación a propiedades del objeto
function mapStateToProps(state) {
    //Devuelvo estado de la aplicación, que contiene la lista de libros.
    //Si hay cambios en la lista el estado se entera del cambio
    return {
        books: state.books,
        filter_book: state.filter_book,
        elements_x_page: state.elements_x_page,
        current_page: state.current_page
    }
}

//Mapear el despachar acciones de mi aplicación a propiedades del objeto
//Vinculo a una propiedad de mi componente una acción definida
function mapDispatchToProps(dispatch){
    // fetchBooks se saca de la definición de acciones
    // Se define la propiedad fetchBooks a la acción fetchBooks
    // Si se llama igual se puede colapsar {fetchBooks: fetchBooks}
    // queda como {fetchBooks}
    return bindActionCreators({ fetchBooks, totalBooks, elementsXPage, currentPage }, dispatch);
}

// Envolver componente en un container y devolver el container.
export default connect(mapStateToProps, mapDispatchToProps)(BookList);