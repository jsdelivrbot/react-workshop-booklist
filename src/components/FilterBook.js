import React, {Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {filterBooks} from '../actions/index';
import {currentPage} from '../actions/index';

class FilterBook extends Component {

    constructor(props) {
        super(props);

        this.state = {
            term: ''
        };

        this.changeFilterText = this.changeFilterText.bind(this);
    }

    changeFilterText(event) {
        this.setState({
            term: event.target.value
        })
    }

    onClickFilter() {
        this.props.filterBooks(this.state.term);
        this.props.currentPage(1);
    }

    render(){
        return (
            <div className="input-group">
                <input type="text"
                       className="form-control"
                       value={this.state.term}
                       onChange={this.changeFilterText}
                       placeholder="Filter books"/>
                <span className="input-group-btn">
                    <button onClick={() => {this.onClickFilter()}} className="btn btn-primary" type="button">Search</button>
                </span>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {}
}

function mapDispatchToProps(dispatch){
    return bindActionCreators({filterBooks, currentPage}, dispatch);
}

// Envolver componente en un container y devolver el container.
export default connect(mapStateToProps, mapDispatchToProps)(FilterBook);