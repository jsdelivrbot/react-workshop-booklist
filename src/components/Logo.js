import React, { Component } from 'react';

class Logo extends Component {

    render(){
        return (
            <div>
                Això seria una icona
            </div>
        );
    }
}

export default Logo;