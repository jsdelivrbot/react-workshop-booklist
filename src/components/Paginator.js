import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { totalBooks } from '../actions/index';
import { currentPage } from '../actions/index';
import { elementsXPage } from '../actions/index';

class Paginator extends Component
{
    constructor(props) {
        super(props);
        this.totalPag = 0;
        this.props.currentPage(1);
        this.props.elementsXPage(10);
    }

    setTotalPag() {
        this.totalPag = Math.ceil(this.props.total_books/this.props.elements_x_page);
        return this.totalPag;
    }

    addPage() {
        if (this.props.current_page < this.totalPag) {
            this.props.currentPage(this.props.current_page+1);
        }
    }

    subPage() {
        if (this.props.current_page > 1) {
            this.props.currentPage(this.props.current_page-1);
        }
    }

    moveFirstPage()
    {
        this.props.currentPage(1);
    }

    moveLastPage()
    {
        this.props.currentPage(this.totalPag);
    }

    render(){
        return (
            <div>
                <span style={{paddingRight: '0.9em'}}>Total: {this.props.total_books}</span>
                <button onClick={() => {this.moveFirstPage()}} className="btn btn-primary" type="button">{'<<'}</button>
                <button onClick={() => {this.subPage()}} className="btn btn-primary" type="button">{'<'}</button>
                <span style={{paddingLeft:'0.5em', paddingRight:'0.5em'}}>{this.props.current_page} / {this.setTotalPag()}</span>
                <button onClick={() => {this.addPage()}} className="btn btn-primary" type="button">{'>'}</button>
                <button onClick={() => {this.moveLastPage()}} className="btn btn-primary" type="button">{'>>'}</button>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {
        total_books: state.total_books,
        current_page: state.current_page,
        elements_x_page: state.elements_x_page
    }
}

function mapDispatchToProps(dispatch){
    return bindActionCreators({ totalBooks, currentPage, elementsXPage }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(Paginator);