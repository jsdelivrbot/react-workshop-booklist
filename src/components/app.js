import React, { Component } from 'react';
import BookList from './BookList';

export default class App extends Component {
  render() {
    return (
      <div className="row">
        <div className="col-md-3">
          <BookList />
        </div>
      </div>
    );
  }
}
