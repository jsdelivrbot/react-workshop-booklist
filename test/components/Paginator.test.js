import {Paginator} from '../../src/components/Paginator';
import React from 'react';
import renderer from 'react-test-renderer';

describe('Paginator component', () => {

    it('It should render an initial paginator', () => {

        let props = {
            total_books: 300,
            current_page: 1,
            elements_x_page: 10,
            totalBooks: jest.fn(),
            currentPage: jest.fn(),
            elementsXPage: jest.fn()
        };


        let tree = renderer.create(
            <Paginator {...props} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

});