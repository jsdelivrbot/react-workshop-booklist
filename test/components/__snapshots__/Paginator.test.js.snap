// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component It should render an initial paginator 1`] = `
<div>
  <span
    style={
      Object {
        "paddingRight": "0.9em",
      }
    }
  >
    Total: 
    300
  </span>
  <button
    className="btn btn-primary"
    onClick={[Function]}
    type="button"
  >
    &lt;&lt;
  </button>
  <button
    className="btn btn-primary"
    onClick={[Function]}
    type="button"
  >
    &lt;
  </button>
  <span
    style={
      Object {
        "paddingLeft": "0.5em",
        "paddingRight": "0.5em",
      }
    }
  >
    1
     / 
    30
  </span>
  <button
    className="btn btn-primary"
    onClick={[Function]}
    type="button"
  >
    &gt;
  </button>
  <button
    className="btn btn-primary"
    onClick={[Function]}
    type="button"
  >
    &gt;&gt;
  </button>
</div>
`;
